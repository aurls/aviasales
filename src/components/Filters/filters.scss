@import '../../assets/styles/styleConstants';

.filters {
  box-shadow: 0 2px 8px rgba(0, 0, 0, .1);
  border-radius: 5px;
  padding: 20px;
  background: white;
}

.filters__title {
  font-size: 12px;
  line-height: 12px;
  font-weight: 600;
  letter-spacing: .5px;
  text-transform: uppercase;
}

.filter {
  display: flex;
  align-items: center;
  height: 40px;
  margin: 0 -20px 0 -20px;
  padding: 0 20px 0 20px;
  cursor: pointer;
  font-size: 13px;
  line-height: 20px;
  user-select: none;

  &:first-child {
    margin-top: 10px;
  }

  &:last-child {
    margin-bottom: -10px;
  }

  &:hover {
    background: $lightPrimaryColor;
    transition: all $standardTransition;
  }
}

.filter__checkbox {
  position: relative;
  width: 20px;
  height: 20px;
  margin: 0 10px 0 0;
  border: 1px solid #9abbce;
  border-radius: 2px;
  padding: 0;
  appearance: none;
  cursor: inherit;
  outline: none;

  &:checked {
    border-color: #2196f3;

    &::before {
      content: "";
      position: absolute;
      top: 5px;
      left: 3px;
      display: block;
      width: 12px;
      height: 8px;
      background: url("./check.png");
      background-size: 100%;
    }
  }
}
