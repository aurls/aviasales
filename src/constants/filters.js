export const ALL = {
  type: 'ALL',
  name: 'Все',
  callback: null
};

export const NO_TRANSFERS = {
  type: 'NO_TRANSFERS',
  name: 'Без пересадок',
  callback: (numberOfTransfers) => numberOfTransfers === 0
};

export const ONE_TRANSFER = {
  type: 'ONE_TRANSFER',
  name: '1 пересадка',
  callback: (numberOfTransfers) => numberOfTransfers === 1
};

export const TWO_TRANSFERS = {
  type: 'TWO_TRANSFERS',
  name: '2 пересадки',
  callback: (numberOfTransfers) => numberOfTransfers === 2
};

export const THREE_TRANSFERS = {
  type: 'THREE_TRANSFERS',
  name: '3 пересадки',
  callback: (numberOfTransfers) => numberOfTransfers === 3
};
