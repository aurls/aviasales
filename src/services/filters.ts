import Types from '../types';

const ALL: Types.Filter = {
  type: 'ALL',
  title: 'Все',
  fn: () => true
};

const NO_TRANSFERS: Types.Filter = {
  type: 'NO_TRANSFERS',
  title: 'Без пересадок',
  fn: (numberOfTransfers: number) => numberOfTransfers === 0
};

const ONE_TRANSFER: Types.Filter = {
  type: 'ONE_TRANSFER',
  title: '1 пересадка',
  fn: (numberOfTransfers: number) => numberOfTransfers === 1
};

const TWO_TRANSFERS: Types.Filter = {
  type: 'TWO_TRANSFERS',
  title: '2 пересадки',
  fn: (numberOfTransfers: number) => numberOfTransfers === 2
};

const THREE_TRANSFERS: Types.Filter = {
  type: 'THREE_TRANSFERS',
  title: '3 пересадки',
  fn: (numberOfTransfers: number) => numberOfTransfers === 3
};

export default {
  ALL,
  NO_TRANSFERS,
  ONE_TRANSFER,
  TWO_TRANSFERS,
  THREE_TRANSFERS
};
