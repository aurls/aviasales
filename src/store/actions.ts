import Types from '../types';
import { Actions } from './storeTypes';
import actionTypes from './actionTypes';

const fetchTicketsRequest = (): Actions.FetchTicketsRequest => ({
  type: actionTypes.FETCH_TICKETS_REQUEST
});

const fetchTicketsSuccess = (tickets: Types.Tickets): Actions.FetchTicketsSuccess => ({
  type: actionTypes.FETCH_TICKETS_SUCCESS,
  payload: tickets
});

const fetchTicketsFailure = (): Actions.FetchTicketsFailure => ({
  type: actionTypes.FETCH_TICKETS_FAILURE
});

const setFilter = (filter: Types.Filter): Actions.SetFilter => ({
  type: actionTypes.SET_FILTER,
  payload: filter
});

const setSorting = (sorting: string): Actions.SetSorting => ({
  type: actionTypes.SET_SORTING,
  payload: sorting
});

export default {
  fetchTicketsRequest,
  fetchTicketsSuccess,
  fetchTicketsFailure,
  setFilter,
  setSorting
};
