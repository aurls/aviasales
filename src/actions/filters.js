import * as actionTypes from '../constants/actionTypes';

export const setFilter = (filter) => {
  return {
    type: actionTypes.SET_FILTER,
    payload: filter
  };
};
