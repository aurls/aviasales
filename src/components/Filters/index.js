import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import * as filters from '../../constants/filters';
import { filters as filtersActions } from '../../actions';
import './filters.scss';

const Filters = ({ appliedFilters, setFilter }) => {
  const isFilterChecked = (filter) => {
    return appliedFilters.includes(filters.ALL.type) || appliedFilters.includes(filter.type);
  };

  const onChangeFilter = (event) => {
    setFilter(event.target.value);
  };

  return (
    <div className="filters">
      <h2 className="filters__title">Количество пересадок</h2>
      <ul>
        {
          Object.values(filters).map((filter) =>
            <li key={filter.type}>
              <label className="filter"
                htmlFor={`filter-${filter.type}`}>
                <input className="filter__checkbox"
                  id={`filter-${filter.type}`}
                  type="checkbox"
                  name="filter"
                  value={filter.type}
                  checked={isFilterChecked(filter)}
                  onChange={onChangeFilter} />
                {filter.name}
              </label>
            </li>)
        }
      </ul>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    appliedFilters: state.appliedFilters
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setFilter: bindActionCreators(filtersActions.setFilter, dispatch)
  };
};

Filters.propTypes = {
  appliedFilters: PropTypes.array.isRequired,
  setFilter: PropTypes.func.isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Filters);
