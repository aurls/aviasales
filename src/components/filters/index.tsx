import React from 'react';
import Types from '../../types';
import actions from '../../store/actions';
import filters from '../../services/filters';
import './filters.scss';

type Props = {
  currentFilters: Types.Filter[]
  setFilter: typeof actions.setFilter
}

const Filters: React.FC<Props> = (props: Props) => {
  const {
    currentFilters,
    setFilter
  } = props;

  const isFilterChecked = (filter: Types.Filter) => {
    return currentFilters.includes(filter) || currentFilters.includes(filters.ALL);
  };

  const onFilterClick = (filter: Types.Filter) => {
    setFilter(filter);
  };

  return (
    <div className="filters">
      <h2 className="filters__title">Количество пересадок</h2>
      <ul>
        {
          Object.values(filters).map((filter) =>
            <li key={filter.type}>
              <label className="filter">
                <input className="filter__checkbox"
                  type="checkbox"
                  name="filter"
                  value={filter.type}
                  checked={isFilterChecked(filter)}
                  onChange={() => onFilterClick(filter)} />
                {filter.title}
              </label>
            </li>)
        }
      </ul>
    </div>
  );
};

export default React.memo(Filters);
