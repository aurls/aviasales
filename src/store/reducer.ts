import Types from '../types';
import actionTypes from './actionTypes';
import { Actions, State } from './storeTypes';
import filters from '../services/filters';
import sortings from '../services/sortings';

const initialState: State = {
  tickets: {},
  isFetching: true,
  hasError: false,
  filters: [filters.ALL],
  sorting: sortings.BY_PRICE_ASC
};

const reducer = (state: State = initialState, action: Actions.All): State => {
  switch (action.type) {
    case actionTypes.FETCH_TICKETS_REQUEST:
      return fetchTicketsRequest(state);

    case actionTypes.FETCH_TICKETS_SUCCESS:
      return fetchTicketsSuccess(state, action.payload);

    case actionTypes.FETCH_TICKETS_FAILURE:
      return fetchTicketsFailure(state);

    case actionTypes.SET_FILTER:
      return setFilter(state, action.payload);

    case actionTypes.SET_SORTING:
      return setSorting(state, action.payload);

    default:
      return state;
  }
};

function fetchTicketsRequest (state: State): State {
  return {
    ...state,
    isFetching: true,
    hasError: false
  };
}

function fetchTicketsSuccess (state: State, newTickets: Types.Tickets): State {
  return {
    ...state,
    tickets: {
      ...state.tickets,
      ...newTickets
    },
    isFetching: false,
    hasError: false
  };
}

function fetchTicketsFailure (state: State): State {
  return {
    ...state,
    isFetching: false,
    hasError: true
  };
}

function setFilter (state: State, filter: Types.Filter): State {
  const updateFilters = (filters: Types.Filter[]) => ({
    ...state,
    filters
  });
  const currentFilters: Types.Filter[] = [...state.filters];

  if (filter === filters.ALL) {
    if (currentFilters.includes(filters.ALL)) {
      return updateFilters([]);
    }
    return updateFilters([filters.ALL]);
  }

  if (currentFilters.includes(filter)) {
    return updateFilters(
      [...currentFilters.filter((i) => i !== filter)]);
  }

  const filtersCount = Object.values(filters).length;
  const currentFiltersCount = Object.values(currentFilters).length;

  if (filtersCount - currentFiltersCount > 2) {
    return updateFilters(
      [...currentFilters.filter((i) => i !== filters.ALL), filter]);
  }

  return updateFilters([filters.ALL]);
}

function setSorting (state: State, sorting: Types.Sorting): State {
  return {
    ...state,
    sorting
  };
}

export default reducer;
