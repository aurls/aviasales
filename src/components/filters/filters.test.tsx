import React from 'react';
import Enzyme from 'enzyme';
import Filters from './index';
import { initialState } from '../../store/reducer';
import filters from '../../services/filters';

describe('Filters component', () => {
  const setFilter = jest.fn();
  let component: Enzyme.ShallowWrapper;

  beforeEach(() => {
    component = Enzyme.shallow(
      <Filters
        currentFilters={initialState.filters}
        setFilter={setFilter} />
    );
  });

  it('Default state', () => {
    expect(component).toMatchSnapshot();
  });

  it('Filter changed', () => {
    const selector = '.filter__checkbox[value="ALL"]';
    component.find(selector).simulate('change');

    expect(setFilter).toHaveBeenCalledWith(filters.ALL);
  });
});
