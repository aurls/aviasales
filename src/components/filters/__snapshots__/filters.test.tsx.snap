// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component Default state 1`] = `
<div
  className="filters"
>
  <h2
    className="filters__title"
  >
    Количество пересадок
  </h2>
  <ul>
    <li
      key="ALL"
    >
      <label
        className="filter"
      >
        <input
          checked={true}
          className="filter__checkbox"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="ALL"
        />
        Все
      </label>
    </li>
    <li
      key="NO_TRANSFERS"
    >
      <label
        className="filter"
      >
        <input
          checked={true}
          className="filter__checkbox"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="NO_TRANSFERS"
        />
        Без пересадок
      </label>
    </li>
    <li
      key="ONE_TRANSFER"
    >
      <label
        className="filter"
      >
        <input
          checked={true}
          className="filter__checkbox"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="ONE_TRANSFER"
        />
        1 пересадка
      </label>
    </li>
    <li
      key="TWO_TRANSFERS"
    >
      <label
        className="filter"
      >
        <input
          checked={true}
          className="filter__checkbox"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="TWO_TRANSFERS"
        />
        2 пересадки
      </label>
    </li>
    <li
      key="THREE_TRANSFERS"
    >
      <label
        className="filter"
      >
        <input
          checked={true}
          className="filter__checkbox"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="THREE_TRANSFERS"
        />
        3 пересадки
      </label>
    </li>
  </ul>
</div>
`;
